File: svc-event/src/main/java/com/beancounter/event/service/EventService.java
package com.beancounter.event.service;

import com.beancounter.common.contracts.PortfoliosResponse;
import com.beancounter.common.event.CorporateEvent;
import com.beancounter.common.input.TrustedEventInput;
import com.beancounter.common.input.TrustedTrnEvent;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.utils.KeyGenUtils;
import com.beancounter.event.contract.CorporateEventResponse;
import com.beancounter.event.contract.CorporateEventsResponse;
import com.beancounter.event.integration.EventPublisher;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
@Slf4j
public class EventService {

  private final EventRepository eventRepository;
  private final PositionService positionService;
  private EventPublisher eventPublisher;

  public EventService(PositionService positionService,
                      EventRepository eventRepository) {
    this.positionService = positionService;
    this.eventRepository = eventRepository;
  }

  @Autowired(required = false)
  public void setEventPublisher(EventPublisher eventPublisher) {
    this.eventPublisher = eventPublisher;
  }


  public Collection<TrustedTrnEvent> processMessage(TrustedEventInput eventRequest) {
    if (eventRequest.getData() != null) {
      return processMessage(
          save(eventRequest.getData())
      );
    }
    return null;
  }

  public Collection<TrustedTrnEvent> processMessage(CorporateEvent event) {
    Collection<TrustedTrnEvent> results = new ArrayList<>();
    PortfoliosResponse response = positionService.findWhereHeld(
        event.getAssetId(),
        event.getRecordDate());
    if (response != null && response.getData() != null) {
      for (Portfolio portfolio : response.getData()) {
        TrustedTrnEvent trnEvent = positionService.process(portfolio, event);
        // Don't create forward dated transactions
        if (trnEvent != null && trnEvent.getTrnInput() != null) {
          if (eventPublisher != null) {
            eventPublisher.send(trnEvent);
          }
          results.add(trnEvent);
        }

      }
    }
    return results;
  }

  public CorporateEvent save(CorporateEvent event) {

    Optional<CorporateEvent> existing = eventRepository.findByAssetIdAndRecordDate(
        event.getAssetId(),
        event.getRecordDate());
    if (existing.isPresent()) {
      return existing.get();
    }

    event.setId(KeyGenUtils.getId());
    return eventRepository.save(event);
  }

  public CorporateEventResponse get(String id) {
    Optional<CorporateEvent> result = eventRepository.findById(id);
    return result
        .map(corporateEvent -> CorporateEventResponse.builder().data(corporateEvent).build())
        .orElse(null);
  }

  public CorporateEventsResponse getAssetEvents(String assetId) {
    Collection<CorporateEvent> events = eventRepository.findByAssetId(assetId);
    CorporateEventsResponse response = CorporateEventsResponse.builder().build();

    for (CorporateEvent event : events) {
      response.getData().add(event);
    }
    return response;

  }

  public Collection<CorporateEvent> forAsset(String assetId) {
    return eventRepository.findByAssetId(assetId);
  }

  public void backFillEvents(String portfolioCode, String valuationDate) {
    positionService.backFillEvents(portfolioCode, valuationDate);
  }

  public Collection<CorporateEvent> findInRange(LocalDate start, LocalDate end) {
    return eventRepository.findByDateRange(start, end);
  }
}

File: svc-event/src/main/java/com/beancounter/event/service/EventService.kt
package com.beancounter.event.service

import com.beancounter.common.event.CorporateEvent
import com.beancounter.common.exception.BusinessException
import com.beancounter.common.input.TrustedEventInput
import com.beancounter.common.input.TrustedTrnEvent
import com.beancounter.common.utils.KeyGenUtils
import com.beancounter.event.contract.CorporateEventResponse
import com.beancounter.event.contract.CorporateEventsResponse
import com.beancounter.event.integration.EventPublisher
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.ArrayList

/**
 * Corporate Action Event Processing Service. Stores and emits events.
 */
@Service
class EventService(
    private val positionService: PositionService,
    private val eventRepository: EventRepository,
    private val keyGenUtils: KeyGenUtils,
) {
    private var eventPublisher: EventPublisher? = null

    @Autowired(required = false)
    fun setEventPublisher(eventPublisher: EventPublisher?) {
        this.eventPublisher = eventPublisher
    }

    fun processMessage(eventRequest: TrustedEventInput): Collection<TrustedTrnEvent> {
        return processMessage(
            save(eventRequest.data)
        )
    }

    fun processMessage(event: CorporateEvent): Collection<TrustedTrnEvent> {
        val results: MutableCollection<TrustedTrnEvent> = ArrayList()
        val response = positionService.findWhereHeld(
            event.assetId,
            event.recordDate
        )
        for (portfolio in response.data) {
            val trnEvent = positionService.process(portfolio, event)
            // Don't create forward dated transactions
            if (trnEvent != null) {
                trnEvent.trnInput
                if (eventPublisher != null) {
                    eventPublisher!!.send(trnEvent)
                }
                results.add(trnEvent)
            }
        }
        return results
    }

    fun save(event: CorporateEvent): CorporateEvent {
        val existing = eventRepository.findByAssetIdAndRecordDate(
            event.assetId,
            event.recordDate
        )
        if (existing.isPresent) {
            return existing.get()
        }
        val save = CorporateEvent(
            keyGenUtils.id,
            event.trnType,
            event.source,
            event.assetId,
            event.recordDate,
            event.rate,
            event.split,
            event.payDate
        )
        return eventRepository.save(save)
    }

    operator fun get(id: String): CorporateEventResponse {
        val result = eventRepository.findById(id)
        return result
            .map { data: CorporateEvent? -> CorporateEventResponse(data!!) }
            .orElseThrow { BusinessException("Not found $id") }
    }

    fun getAssetEvents(assetId: String): CorporateEventsResponse {
        val events = eventRepository.findByAssetId(assetId)
        if (events.isNullOrEmpty()) {
            return CorporateEventsResponse()
        }
        return CorporateEventsResponse(events)
    }

    fun forAsset(assetId: String): Collection<CorporateEvent> {
        return eventRepository.findByAssetId(assetId)
    }

    fun backFillEvents(portfolioId: String, valuationDate: String) {
        positionService.backFillEvents(portfolioId, valuationDate)
    }

    fun findInRange(start: LocalDate, end: LocalDate): Collection<CorporateEvent> {
        return eventRepository.findByDateRange(start, end)
    }

    fun getScheduledEvents(start: LocalDate): CorporateEventsResponse {
        val events = eventRepository.findByStartDate(start)
        return CorporateEventsResponse(events)
    }
}
